src/iocp/mod.rs

```diff
@@ -323,6 +323,11 @@ impl Poller {
         self.port.post(0, 0, self.notifier.clone())
     }
 
+    /// Push an IOCP packet into the queue.
+    pub(super) fn post(&self, packet: &CompletionPacket) -> io::Result<()> {
+        self.port.post(0, 0, packet.clone().0)
+    }
```

Comment:
why do we clone the packet (and don't pass it as is)? It doesn't have a special `Clone` or `Drop` implementation, so this shouldn't be necessary.

Reply:
`Packet` is a typedef for `Pin<Arc<...>>`, so that the underlying packet internals aren't dropped while the packet is waiting in the AFD system for an event. `IoCompletionPort::post()` expects this `Pin<Arc<...>>` for this reason.

Reply:
how is this expected to be used?

Reply:
The general idea that it can be used for events that are out of band of the `Poller`, for instance:

```rust
let poller = Arc::new(Poller::new().unwrap());
let my_source: TcpStream = /* ... */;
#[cfg(unix)]
let external_source: EventFd = /* ... */;
let my_key: usize = 0;
let external_key: usize = 1;

// Register the sources.
poller.add(my_source, Event::readable(my_key));
#[cfg(unix)]
poller.add(external_source, Event::readable(external_key));

// Wait for something to happen.
thread::spawn({
    let poller = poller.clone();
    move || {
        #[cfg(windows)]
        let packet = CompletionPacket::new(Event::readable(external_key));

        wait_for_something_blocking_to_happen();
        #[cfg(unix)]
        // On unix, we can use pipes/eventfds, like so:
        external_source.signal();

        #[cfg(windows)]
        // On windows, we don't have an equivalent to eventfd/pipe, so we need to
        // post the event
        poller.post(&packet);
    }
});

loop {
    let mut events = vec![];
    poller.wait(&mut events, None)?;

    for event in events {
        if event.key == my_key {
            handle_my_source(&my_source)?;
        } else if event.key == external_key {
            handle_external_source()?;
        }
    }
}
```

The motivating use case is to be able to shove Windows anonymous pipes into another thread but still be able to wait on them. See [here](https://github.com/alacritty/alacritty/pull/6846#issuecomment-1500933116) for more info.

Reply:
Realized that you might be asking why I took this API design pattern instead of just taking an `Event`.

Imagine that, instead of the above `thread::spawn()` block being a one-off wait, that it's instead a loop that calls `wait_for_something_blocking_to_happen()`, then calling the `post()` function. If the `post()` function just took an `Event`, it would have to re-allocate the `Packet` every time it posted an event. With this strategy, no extra allocation has to occur.